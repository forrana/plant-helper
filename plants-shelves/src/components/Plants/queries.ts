import { gql } from '@apollo/client';
const GET_ALL_PLANTS = gql`
query {
  plants {
    id, name, scientificName, daysUntilNextWatering, daysBetweenWatering, daysBetweenWateringGrowing, daysBetweenWateringDormant, daysPostpone
    symbol {
      userWideId
    }
    room {
      id,
      roomName,
      colorBackground
    }
  }
}
`;

const GET_ALL_ROOMS = gql`
  query {
    rooms {
      id, roomName, colorBackground
    }
  }
`;
const ROOMS_BY_NAME_FRAGMENT = gql`
  query RoomsByNameFragment($nameFragment: String!) {
    roomsByNameFragment(nameFragment: $nameFragment) {
      id, roomName, colorBackground
    }
  }
`;

const ADD_PLANT = gql`
  mutation CreatePlant($plantName: String!, $scientificName: String!, $daysBetweenWateringGrowing: Int!, $daysBetweenWateringDormant: Int!,
    $groupName: String, $colorBackground: String) {
      createPlant(plantName: $plantName, scientificName: $scientificName, daysBetweenWateringGrowing: $daysBetweenWateringGrowing,
        daysBetweenWateringDormant: $daysBetweenWateringDormant, groupName: $groupName, colorBackground: $colorBackground) {
        plant {
          id, name, scientificName, daysUntilNextWatering, daysBetweenWatering, daysBetweenWateringGrowing, daysBetweenWateringDormant, daysPostpone
          symbol {
            userWideId
          }
          room {
            id,
            roomName,
            colorBackground
          }
        }
      }
  }
`;

const WATER_PLANT = gql`
  mutation WaterPlant($plantId: ID!) {
    waterPlant(plantId: $plantId) {
      plant {
        id, name, scientificName, daysUntilNextWatering, daysBetweenWatering, daysBetweenWateringGrowing, daysBetweenWateringDormant, daysPostpone
        symbol {
          userWideId
        }
        room {
          id,
          roomName,
          colorBackground
        }
      }
    }
  }
`;

const POSTPONE_WATERING = gql`
  mutation PostponeWatering($plantId: ID!, $postponeDays: Int!) {
    postponeWatering(plantId: $plantId, days: $postponeDays) {
      plant {
        id, name, scientificName, daysUntilNextWatering, daysBetweenWatering, daysBetweenWateringGrowing, daysBetweenWateringDormant, daysPostpone
        symbol {
          userWideId
        }
        room {
          id,
          roomName,
          colorBackground
        }
      }
    }
  }
`;

const UPDATE_PLANT = gql`
  mutation UpdatePlant($plantId: ID!, $plantName: String!,
    $scientificName: String!, $daysBetweenWateringGrowing: Int!, $daysBetweenWateringDormant: Int!,
    $postponeDays: Int!, $groupName: String, $colorBackground: String) {
    updatePlant(plantId: $plantId, plantName: $plantName, scientificName: $scientificName,
      daysBetweenWateringGrowing: $daysBetweenWateringGrowing,
      daysBetweenWateringDormant: $daysBetweenWateringDormant,
      postponeDays: $postponeDays,
      groupName: $groupName,
      colorBackground: $colorBackground) {
      plant {
        id, name, scientificName, daysUntilNextWatering,
        daysBetweenWatering, daysBetweenWateringGrowing, daysBetweenWateringDormant,
        daysPostpone,
        symbol {
          userWideId
        }
        room {
          id,
          roomName,
          colorBackground
        }
      }
    }
  }
`;

const UPDATE_ROOM = gql`
  mutation UpdateRoom($roomId: ID!, $roomName: String!) {
    updateRoom(roomId: $roomId, roomName: $roomName) {
      ok
      room {
        id,
        roomName,
        colorBackground
      }
    }
  }
`

const CREATE_ROOM = gql`
  mutation CreateRoom($plant1Id: ID!, $plant2Id: ID!) {
    createRoom(plant1Id: $plant1Id, plant2Id: $plant2Id) {
      ok
    }
  }
`;

const DELETE_PLANT = gql`
  mutation DeletePlant($plantId: ID!) {
    deletePlant(plantId: $plantId) {
      ok
    }
  }
`;

const GET_SUBSCRIPTION = gql`
  query {
    subscription
  }
`

const CREATE_SUBSCRIPTION = gql`
  mutation CreateSubscription($endpoint: String!, $p256dh: String!, $auth: String!, $permissionGiven: Boolean!) {
    createSubscription(endpoint: $endpoint, p256dh: $p256dh, auth: $auth, permissionGiven: $permissionGiven) {
      ok
    }
  }
`

export {
  GET_ALL_PLANTS,
  GET_ALL_ROOMS,
  ADD_PLANT,
  WATER_PLANT,
  UPDATE_PLANT,
  DELETE_PLANT,
  CREATE_SUBSCRIPTION,
  GET_SUBSCRIPTION,
  POSTPONE_WATERING,
  CREATE_ROOM,
  ROOMS_BY_NAME_FRAGMENT,
  UPDATE_ROOM
}
