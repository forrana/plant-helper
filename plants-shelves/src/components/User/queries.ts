import { gql } from '@apollo/client';

const CREATE_USER = gql`
  mutation Register($email: String!, $username: String!, $password1: String!, $password2: String!) {
    register(email: $email, username: $username, password1: $password1, password2: $password2) {
      success,
      errors
    }
  }
`;

const LOG_IN = gql`
  mutation TokenAuth($username: String!, $password: String!) {
    tokenAuth(username: $username, password: $password) {
      success,
      errors,
      unarchiving,
      token,
      refreshToken,
      unarchiving,
      user {
        id,
        username,
      }
    }
  }
`;


const REFRESH_TOKEN = gql`
  mutation RefreshToken($refreshToken: String!) {
    refreshToken(
      refreshToken: $refreshToken
    ) {
      success,
      errors,
      payload,
      token,
      refreshToken,
    }
  }
`

const UPSERT_USER_SETTINGS = gql`
  mutation UpsertUserSettings($startTime: String!, $endTime: String!, $timezone: String) {
    upsertUserSettings(startTime: $startTime, endTime: $endTime, timezone: $timezone) {
      ok
    }
  }
`

const GET_USER_SETTINGS = gql`
  query {
    userSettings {
      notificationsStartTime, notificationsEndTime, timezone
    }
  }
`;

export { CREATE_USER, LOG_IN, REFRESH_TOKEN, UPSERT_USER_SETTINGS, GET_USER_SETTINGS }
