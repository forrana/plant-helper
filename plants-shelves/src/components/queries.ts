import { gql } from '@apollo/client';
const GET_ALL_PLANTS = gql`
query {
  plants {
    id, name, scientificName, daysUntilNextWatering, daysBetweenWatering
  }
}
`;

const ADD_PLANT = gql`
  mutation CreatePlant($plantName: String!, $scientificName: String!) {
    createPlant(plantName: $plantName, scientificName: $scientificName) {
      plant {
        name,
        scientificName
      }
    }
  }
`;

const WATER_PLANT = gql`
  mutation WaterPlant($plantId: ID!) {
    waterPlant(plantId: $plantId) {
      plant {
        id, name, scientificName, daysUntilNextWatering, daysBetweenWatering
      }
    }
  }
`;


export { GET_ALL_PLANTS, ADD_PLANT, WATER_PLANT }
