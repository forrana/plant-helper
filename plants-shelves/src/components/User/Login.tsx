import React, { useState } from 'react';
import { Button, Form, FormGroup, Label, Input, Spinner, FormFeedback } from 'reactstrap';
import { useMutation } from '@apollo/client';
import { Link } from "react-router-dom";
import { LOG_IN } from './queries'
import styles from "./Login.module.css"
import { LoginError } from "./models"

function Login() {
  const [login, setLogin] = useState("");
  const [password, setPassword] = useState("");
  const [loginErrors, setLoginErrors] = useState<Array<LoginError>>([]);

  const [loginUser, { loading, error }] = useMutation(LOG_IN, {
    onCompleted: (data: any) => {
      setLogin("");
      setPassword("");
      const errors: Array<LoginError> = data?.tokenAuth?.errors?.nonFieldErrors;
      if(errors) {
        setLoginErrors(errors);
      } else setLoginErrors([]);
    }
  });

  const handleFormSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    if(login && password) {
      loginUser({ variables: { username: login, password }})
    }
  }

  const handleLoginInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setLogin(event.target.value);
  };

  const handlePasswordInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setPassword(event.target.value);
  };

  if (loading) return  <Spinner color="primary" />

  if (error)  return  <p>Error :( {error.message}</p>;

  return (
    <Form
      onSubmit={handleFormSubmit}
      autoComplete="off"
      className={styles.container}
    >
      <FormGroup>
        <Label for="login">Login:</Label>
        <Input type="text" name="login" id="login" placeholder="Enter login"
          value={login}
          onChange={handleLoginInputChange}
          invalid={loginErrors.length > 0}
          required
        />
        {
          loginErrors.map((error, index) =>
            <FormFeedback key={index}>{ error.message }</FormFeedback>
          )
        }
      </FormGroup>
      <FormGroup>
        <Label for="password">Password:</Label>
        <Input type="password" name="password" id="password" placeholder="Enter password"
          value={password}
          onChange={handlePasswordInputChange}
          invalid={loginErrors.length > 0}
          required
        />
        {
          loginErrors.map((error, index) =>
            <FormFeedback key={index}>{ error.message }</FormFeedback>
          )
        }
      </FormGroup>
      <section className={styles.controls}>
        <Button type="submit">Login</Button>
        <Link to="/signup" className={styles.link}>Signup</Link>
      </section>
    </Form>
  )
}

export default Login
