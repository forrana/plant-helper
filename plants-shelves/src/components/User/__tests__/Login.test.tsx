import React from 'react';
import { act, fireEvent, render, screen, Screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';

import { MemoryRouter } from 'react-router-dom';
import Login from '../Login';
import { LOG_IN } from '../queries';

const user = {
  username: "username",
  password: "password" ,
  token: "token"
 };

const mocksWithUser: any = [
  {
    request: {
      query: LOG_IN,
      variables: { username: user.username, password: user.password }
    },
    result: {
      data: {
        tokenAuth: {
          user: user,
          token: user.token
        }
      }
    },
  }
];

const mockCallback = jest.fn(() => null)

test('Match snapshot', async () => {
  const {container} = render(
    <MemoryRouter>
      <MockedProvider mocks={mocksWithUser} addTypename={false}>
        <Login />
      </MockedProvider>
    </MemoryRouter>,
  );

  expect(container).toMatchSnapshot()
});

test('Allow to login', async () => {
 const { container } = render(
    <MemoryRouter>
      <MockedProvider mocks={mocksWithUser} addTypename={false}>
      <Login />
      </MockedProvider>
    </MemoryRouter>,
  );

  await act(async () => {
    fireEvent.change(
      screen.getByLabelText("Login:"),
      { target: { value: user.username } }
    )
  });

  await act(async () => {
    fireEvent.change(
      screen.getByLabelText("Password:"),
      { target: { value: user.password } }
    )
  });

  await act(async () => {
    fireEvent.click(
      screen.getByTestId("login-submit-button")
    )
    await new Promise(resolve => setTimeout(resolve, 0));
  });

  expect(screen.getByLabelText("Password:")).not.toHaveValue(user.password)
})
