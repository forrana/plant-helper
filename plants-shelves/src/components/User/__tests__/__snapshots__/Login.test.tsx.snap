// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Match snapshot 1`] = `
<div>
  <main>
    <form
      autocomplete="off"
      class="container"
    >
      <div
        class="form-group"
      >
        <label
          class=""
          for="login"
        >
          Login:
        </label>
        <input
          aria-invalid="false"
          class="form-control"
          id="login"
          name="login"
          placeholder="Enter login"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class=""
          for="password"
        >
          Password:
        </label>
        <input
          aria-invalid="false"
          class="form-control"
          id="password"
          name="password"
          placeholder="Enter password"
          required=""
          type="password"
          value=""
        />
      </div>
      <section
        class="controls"
      >
        <button
          class="btn btn-secondary"
          data-testid="login-submit-button"
          type="submit"
        >
          Login
        </button>
        <a
          class="link"
          href="/signup"
        >
          Signup
        </a>
      </section>
    </form>
  </main>
</div>
`;
